File: backend/app/tools/web_tools.py
import requests
from bs4 import BeautifulSoup
from tavily import TavilyClient

from app.core.config import TAVILY_API_KEY

tavily = TavilyClient(api_key=TAVILY_API_KEY)

def web_search(query: str) -> str:
    """
    Search the web using Tavily and return formatted results.
    """
    try:
        results = tavily.search(query=query, max_results=5)

        output = []

        for r in results.get("results", []):
            output.append(
                f"Title: {r['title']}\n"
                f"URL: {r['url']}\n"
                f"Snippet: {r['content'][:200]}\n"
            )

        return "\n----\n".join(output)

    except Exception as e:
        return f"Error during web search: {str(e)}"

def scrape_url(url: str) -> str:
    """
    Scrape a webpage and return clean text content.
    """
    try:
        headers = {
            "User-Agent": "Mozilla/5.0"
        }

        response = requests.get(url, timeout=10, headers=headers)

        soup = BeautifulSoup(response.text, "html.parser")

        for tag in soup(["script", "style", "nav", "footer"]):
            tag.decompose()

        text = soup.get_text(separator=" ", strip=True)

        return text[:3000]

    except Exception as e:
        return f"Error scraping URL: {str(e)}"

File: syncai/src/lib/tools/webTools.ts
import * as cheerio from "cheerio";

export async function tavilySearch(query: string): Promise<string> {
  const apiKey = process.env.TAVILY_API_KEY;
  if (!apiKey) throw new Error("TAVILY_API_KEY is not set in .env.local");

  const res = await fetch("https://api.tavily.com/search", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ api_key: apiKey, query, max_results: 5 }),
  });

  if (!res.ok) throw new Error(`Tavily API error: ${res.statusText}`);

  const data = await res.json();
  const results: { title: string; url: string; content: string }[] =
    data.results ?? [];

  return results
    .map(
      (r) =>
        `Title: ${r.title}\nURL: ${r.url}\nSnippet: ${r.content?.slice(0, 300) ?? ""}`,
    )
    .join("\n----\n");
}

export async function scrapeUrl(url: string): Promise<string> {
  const res = await fetch(url, {
    headers: { "User-Agent": "Mozilla/5.0" },
    signal: AbortSignal.timeout(10_000),
  });

  const html = await res.text();
  const $ = cheerio.load(html);
  $("script, style, nav, footer, header, aside").remove();
  return $("body").text().replace(/\s+/g, " ").trim().slice(0, 3000);
}

export function extractFirstUrl(searchResults: string): string | null {
  const match = searchResults.match(/URL:\s*(https?:\/\/[^\n]+)/);
  return match ? match[1].trim() : null;
}
